packages/core/core/src/requests/EntryRequest.js

```diff
@@ -143,7 +165,9 @@ class EntryResolver {
     throw new Error(`Unknown entry ${entry}`);
   }
 
-  async readPackage(entry: FilePath) {
+  // Disabling flow because it's complaining that this function is typed as 'any'... but it really should be typed as 'any'
+  // $FlowFixMe
+  async readPackage(entry: FilePath): any {
```

Comment:
We have a `PackageJSON` type in `@parcel/types`. So maybe `{|...PackageJSON, filePath: FilePath|}` works

Reply:
I'm a Typescript pro but a noob at Flow. Can you pull and take a look at that one. I tried this and some similar types, but then it pushes type problems up to the readPackage call which are beyond my skill to resolve.

`PackageJSON & {| filePath: FilePath, source?: string |} | null`

Reply:
Do this:

```diff
diff --git packages/core/core/src/requests/EntryRequest.js packages/core/core/src/requests/EntryRequest.js
index 6692b71f..8a31bb5d 100644
--- packages/core/core/src/requests/EntryRequest.js
+++ packages/core/core/src/requests/EntryRequest.js
@@ -1,6 +1,6 @@
 // @flow strict-local
 
-import type {Async, FilePath, File} from '@parcel/types';
+import type {Async, FilePath, File, PackageJSON} from '@parcel/types';
 import type {StaticRunOpts} from '../RequestTracker';
 import type {Entry, ParcelOptions} from '../types';
 
@@ -97,12 +97,13 @@ export class EntryResolver {
     if (stat.isDirectory()) {
       let pkg = await this.readPackage(entry);
       if (pkg && typeof pkg.source === 'string') {
-        let source = path.join(path.dirname(pkg.filePath), pkg.source);
+        let pkgSource = pkg.source;
+        let source = path.join(path.dirname(pkg.filePath), pkgSource);
         try {
           stat = await this.options.inputFS.stat(source);
         } catch (err) {
           throw new Error(
-            `${pkg.source} in ${path.relative(
+            `${pkgSource} in ${path.relative(
               this.options.inputFS.cwd(),
               pkg.filePath,
             )}#source does not exist`,
@@ -111,7 +112,7 @@ export class EntryResolver {
 
         if (!stat.isFile()) {
           throw new Error(
-            `${pkg.source} in ${path.relative(
+            `${pkgSource} in ${path.relative(
               this.options.inputFS.cwd(),
               pkg.filePath,
             )}#source is not a file`,
@@ -124,7 +125,7 @@ export class EntryResolver {
         if (pkg.targets) {
           for (let targetName in pkg.targets) {
             let target = pkg.targets[targetName];
-            if (target.source) {
+            if (target.source != null) {
               let filePath = path.join(entry, target.source);
               entries.push({filePath, packagePath: entry, target: targetName});
             }
@@ -165,9 +166,9 @@ export class EntryResolver {
     throw new Error(`Unknown entry ${entry}`);
   }
 
-  // Disabling flow because it's complaining that this function is typed as 'any'... but it really should be typed as 'any'
-  // $FlowFixMe
-  async readPackage(entry: FilePath): any {
+  async readPackage(
+    entry: FilePath,
+  ): Promise<?{...PackageJSON, filePath: FilePath, ...}> {
     let content, pkg;
     let pkgFile = path.join(entry, 'package.json');
     try {
```

The problems you ran into are:
1. `path.dirname(pkg.filePath)` is a function call and invalidates the `typeof === "string"` refinement.
2. Async functions need to return `Promise<actual return value>`
1. (For objects in Flow, `{...PackageJSON, something else}` is preferred over `&`.)

Reply:
Thanks, that seems to work.